types/rdf-js/index.d.ts

```diff
@@ -476,3 +476,183 @@ export interface DatasetCoreFactory<Q extends BaseQuad = Quad> {
+     * value of the `iteratee` is used as `accumulator` value for the next calls.
+     *
+     * This method returns the return value of the last `iteratee` call.
+     *
+     * This method is aligned with `Array.prototype.reduce()` in ECMAScript-262.
+     */
+    reduce<A = any>(iteratee: QuadReduceIteratee<Q, A>, initialValue: A): A;
+
+    /**
+     * Existential quantification method, tests whether some quads in the dataset pass the test implemented by the
+     * provided `iteratee`.
+     *
+     * This method immediately returns boolean `true` once a quad that passes the test is found.
+     *
+     * This method is aligned with `Array.prototype.some()` in ECMAScript-262.
+     */
+    some(iteratee: QuadFilterIteratee<Q>): boolean;
+
+    /**
+     * Returns the set of quads within the dataset as a host language native sequence, for example an `Array` in
+     * ECMAScript-262.
+     *
+     * Since a `Dataset` is an unordered set, the order of the quads within the returned sequence is arbitrary.
+     */
+    toArray(): Q[];
+
+    /**
+     * Returns an N-Quads string representation of the dataset, preprocessed with
+     * {@link https://json-ld.github.io/normalization/spec/|RDF Dataset Normalization} algorithm.
+     */
+    toCanonical(): string;
+
+    /**
+     * Returns a stream that contains all quads of the dataset.
+     */
+    toStream(): Stream<Q>;
+
+    /**
+     * Returns an N-Quads string representation of the dataset.
+     *
+     * No prior normalization is required, therefore the results for the same quads may vary depending on the `Dataset`
+     * implementation.
+     */
+    toString(): string;
+
+    /**
+     * Returns a new `Dataset` that is a concatenation of this dataset and the quads given as an argument.
+     */
+    union(quads: Dataset<Q>): Dataset<Q>;
+}
+
+export interface DatasetFactory extends DataFactory {
+    /**
+     * Returns a new dataset and imports all quads, if given.
+     */
+    dataset<Q extends BaseQuad = Quad>(quads?: Dataset<Q>): Dataset<Q>;
+
+    dataset<Q extends BaseQuad = Quad>(quads?: Iterable<Q>): Dataset<Q>;
+}
+
+export interface QuadFilterIteratee<Q extends BaseQuad = Quad> {
+    /**
+     * A callable function that returns `true` if the input quad passes the test this function implements.
+     */
+    test(quad: Q, dataset: Dataset<Q>): boolean;
+}
+
+export interface QuadMapIteratee<Q extends BaseQuad = Quad> {
+    /**
+     * A callable function that can be executed on a quad and returns a quad.
+     *
+     * The returned quad can be the given quad or a new one.
+     */
+    map(quad: Q, dataset: Dataset<Q>): Q;
+}
+
+export interface QuadReduceIteratee<Q extends BaseQuad = Quad, A = any> {
+    /**
+     * A callable function that can be executed on an accumulator and quad and returns a new accumulator.
+     */
+    run(accumulator: A, quad: Q, dataset: Dataset<Q>): A;
```

Comment:
The spec has a `void` return type, as well as saying it 'returns a new accumulator' (the former must be a mistake).

Reply:
Good catch: https://github.com/rdfjs/dataset-spec/pull/56